File: leetcode/C++/map-sum-pairs.cpp
class Trie{
public:
    int n;
    Trie *next;
    Trie(): n(0), next(nullptr) {}
};

class MapSum {
public:
    /** Initialize your data structure here. */
    Trie *node;
    unordered_map<string,int> um;
    MapSum() {node=new Trie();}
    
    void insert(string key, int val) {
        int idx=0, diff=val-um[key];
        um[key]=val;
        Trie* cur=node;
        for(char &ch : key){
            if(cur->next==nullptr) cur->next=new Trie[26];
            cur->next[ch-'a'].n+=diff;
            cur=&(cur->next[ch-'a']);
        }
    }
    
    int sum(string prefix) {
        int sol=0;
        Trie* cur=node;
        for(char &ch : prefix){
            if(cur->next==nullptr || cur->next[ch-'a'].n==0) {
                sol=0;
                break;
            }         
            sol=cur->next[ch-'a'].n;
            cur=&(cur->next[ch-'a']);
        }
        return sol;
    }
};

/*
class MapSum {
public:
    unordered_map<string,int> um, um_str;
    void insert(string key, int val) {
        int diff=val-um_str[key];
        string s;
        um_str[key]=val;
        for(char &ch : key){
            s.push_back(ch);
            um[s]+=diff;
        }
    }
    int sum(string prefix) {
        return um[prefix];
    }
};
 */

File: leetcode/Python/map-sum-pairs.py
class Node:
    def __init__(self):
        self.n = 0
        self.next = dict()

class MapSum:

    def __init__(self):
        self.node = dict()
        self.map = defaultdict(int)

    def insert(self, key: str, val: int) -> None:
        diff, cur = val - self.map[key], self.node
        self.map[key] = val
        for k in key:
            if not k in cur:
                cur[k] = Node()
            cur[k].n+=diff
            cur=cur[k].next

    def sum(self, prefix: str) -> int:
        sol, cur = 0, self.node
        for k in prefix:
            if not k in cur:
                return 0
            sol = cur[k].n
            cur=cur[k].next
        return sol


# Your MapSum object will be instantiated and called as such:
# obj = MapSum()
# obj.insert(key,val)
# param_2 = obj.sum(prefix)

'''
class MapSum:
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.um = defaultdict(int)
        self.um_str = defaultdict(int)

    def insert(self, key: str, val: int) -> None:
        diff = val - self.um_str[key]
        self.um_str[key]=val
        for i in range(len(key)):
            self.um[key[0:i+1]]+=diff

    def sum(self, prefix: str) -> int:
        return self.um[prefix]
'''